subsys/net/lib/wifi_provision/wifi_provision.c

```diff
@@ -0,0 +1,995 @@
+
+	return 0;
+}
+
+static int on_message_complete(struct http_parser *parser)
+{
+	struct http_req *req = CONTAINER_OF(parser, struct http_req, parser);
+
+	req->received_all = true;
+
+	LOG_DBG("on_message_complete, method: %d", parser->method);
+
+	return 0;
+}
+
+static int on_url(struct http_parser *parser, const char *at, size_t length)
+{
+	struct http_req *req = CONTAINER_OF(parser, struct http_req, parser);
+
+	req->url = at;
+	req->url_len = length;
+
+	LOG_DBG("on_url, method: %d", parser->method);
+	LOG_DBG("> %.*s", length, at);
+
+	return 0;
+}
+
+int wifi_provision_start(const wifi_provision_evt_handler_t handler)
+{
+	int ret;
+
+	/* Set library callback handler */
+	handler_cb = handler;
+
+	if (!wifi_credentials_is_empty()) {
+		LOG_DBG("Stored Wi-Fi credentials found, already provisioned");
+		smf_set_initial(SMF_CTX(&state_object), &state[STATE_FINISHED]);
+		return 0;
+	}
+
+	/* Provision self-signed server certificate */
+	ret = tls_credential_add(CONFIG_WIFI_PROVISION_CERTIFICATE_SEC_TAG,
+				 TLS_CREDENTIAL_CA_CERTIFICATE,
+				 server_certificate,
+				 sizeof(server_certificate));
+	if (ret == -EEXIST) {
+		LOG_DBG("CA already exists, sec tag: %d",
+			CONFIG_WIFI_PROVISION_CERTIFICATE_SEC_TAG);
+	} else if (ret < 0) {
+		LOG_ERR("Failed to register CA certificate: %d", ret);
+		return ret;
+	}
+
+	ret = tls_credential_add(CONFIG_WIFI_PROVISION_CERTIFICATE_SEC_TAG,
+				 TLS_CREDENTIAL_SERVER_CERTIFICATE,
+				 server_certificate,
+				 sizeof(server_certificate));
+	if (ret == -EEXIST) {
+		LOG_DBG("Public certificate already exists, sec tag: %d",
+			CONFIG_WIFI_PROVISION_CERTIFICATE_SEC_TAG);
+	} else if (ret < 0) {
+		LOG_ERR("Failed to register public certificate: %d", ret);
+		return ret;
+	}
+
+	ret = tls_credential_add(CONFIG_WIFI_PROVISION_CERTIFICATE_SEC_TAG,
+				 TLS_CREDENTIAL_PRIVATE_KEY,
+				 server_private_key, sizeof(server_private_key));
+
+	if (ret == -EEXIST) {
+		LOG_DBG("Private key already exists, sec tag: %d",
+			CONFIG_WIFI_PROVISION_CERTIFICATE_SEC_TAG);
+	} else if (ret < 0) {
+		LOG_ERR("Failed to register private key: %d", ret);
+		return ret;
+	}
+
+	LOG_DBG("Self-signed server certificate provisioned");
+
+	net_mgmt_init_event_callback(&net_l2_mgmt_cb,
+				     net_mgmt_wifi_event_handler,
+				     NET_MGMT_WIFI);
+
+	net_mgmt_add_event_callback(&net_l2_mgmt_cb);
+
+	smf_set_initial(SMF_CTX(&state_object), &state[STATE_UNPROVISIONED]);
+
+	http_parser_settings_init(&parser_settings);
+
+	parser_settings.on_body = on_body;
+	parser_settings.on_headers_complete = on_headers_complete;
+	parser_settings.on_message_begin = on_message_begin;
+	parser_settings.on_message_complete = on_message_complete;
+	parser_settings.on_url = on_url;
+
+	/* Start processing incoming IPv4 clients */
+	ret = process_tcp4();
+	if (ret < 0) {
+		LOG_ERR("Failed to start TCP server %d", ret);
+		return ret;
+	}
+
+	k_sem_take(&wifi_provision_sem, K_FOREVER);
```

Comment:
Would likely make sense to remove the registered net_mgm handler before leaving, it should no longer need to process events (`net_mgmt_del_event_callback()`).